pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -2078,16 +2232,62 @@ func buildCompleteFilterChain(pluginParams *plugin.InputParams, mutable *plugin.
 	}
 
 	httpConnectionManagers := make([]*http_conn.HttpConnectionManager, len(mutable.FilterChains))
+	thriftProxies := make([]*thrift_proxy.ThriftProxy, len(mutable.FilterChains))
 	for i := range mutable.FilterChains {
 		chain := mutable.FilterChains[i]
 		opt := opts.filterChainOpts[i]
 		mutable.Listener.FilterChains[i].Metadata = opt.metadata
 
-		// we are building a network filter chain (no http connection manager) for this filter chain
-		// In HTTP, we need to have mixer, RBAC, etc. upfront so that they can enforce policies immediately
-		// For network filters such as mysql, mongo, etc., we need the filter codec upfront. Data from this
-		// codec is used by RBAC or mixer later.
-		if opt.httpOpts == nil {
+		if opt.thriftOpts != nil && features.EnableThriftFilter.Get() {
+			var quotas []model.Config
+			// Add the TCP filters first.. and then the Thrift filter
+			mutable.Listener.FilterChains[i].Filters = append(mutable.Listener.FilterChains[i].Filters, chain.TCP...)
+
+			thriftProxies[i] = buildThriftProxy(opt.thriftOpts)
+
+			if pluginParams.Service != nil {
+				quotas = opts.push.IstioConfigStore.QuotaSpecByDestination(&model.ServiceInstance{
+					Service: pluginParams.Service,
```

Comment:
Why do you need this ?

Reply:
This is going to be very slow! You might want to cache this in the push context like other crds.

Reply:
If we don't do this, then the rate limit filter will be applied to all outbound connections. This means that every thrift within the mesh will generate a call to the rate limit service, where we can scope it down to only those connections where it should be applied. I'll look into caching!